src/fifelse.c

```diff
@@ -343,3 +343,194 @@ SEXP fcaseR(SEXP na, SEXP rho, SEXP args) {
   UNPROTECT(nprotect);
   return ans;
 }
+
+SEXP topnR(SEXP vec, SEXP n, SEXP dec) {
+	int nprotect = 0;
+	int64_t i, j, idx = 0;
+	int len0 = asInteger(n);
+	const int64_t len1 = xlength(vec);
+	
+	if (isS4(vec) && !INHERITS(vec, char_nanotime)) {
+		error("S4 class objects (excluding nanotime) are not supported.");
+	}
+	if (len0 > 1000) {
+		error("Function 'topn' is not built for large value of 'n'. The algorithm is made for small values. Please prefer the 'order' function if you want to proceed with such large value.");
+	}
+	if (len0 > len1) {
+		warning("'n' is larger than length of 'vec'. 'n' will be set to length of 'vec'.");
+		len0 = len1;
+	}
+	if (len0 < 1) {
+		error("Please enter a positive integer larger or equal to 1.");
+	}
+	if (!IS_TRUE_OR_FALSE(dec)) {
+		error("Argument 'decreasing' must be TRUE or FALSE and length 1.");
+	}
+	
+	const bool vdec = LOGICAL(dec)[0];
+	SEXPTYPE tvec = TYPEOF(vec);
+	SEXP ans = PROTECT(allocVector(INTSXP, len0)); nprotect++;
+	int *restrict pans = INTEGER(ans);
+	int tmp;
+
+	if (vdec) {
+		switch(tvec) {
+		case INTSXP: {
+			const int *restrict pvec = INTEGER(vec);
+			int min_value = pvec[0];
+			for (i = 0; i < len0; ++i) {
+				pans[i] = i;
+				if (pvec[i] <= min_value || pvec[i] == NA_INTEGER) {
+					min_value = pvec[i];
+					idx = i;
+				}
+			}
+			for (i = len0; i < len1; ++i) {
+				if (pvec[i] == NA_INTEGER) {
+					continue;
+				}
+				if (pvec[i] > min_value) {
+					min_value = pvec[i];
+					pans[idx] = i;
+					for (j = 0; j <len0; ++j) {
```

Comment:
I'll have to read more carefully, but it looks like we do a full re-sort of the `len0` indices each time we find a new candidate is that right? Can we try to compare the min-heap implementation?

Reply:
No, we don't do a full re-sort, we just recompute the min/max value and its position to be able to compare it to future values.

Reply:
The sorting of the final vector is done in the next loop. Maybe I need to add some comments to explain what I did?